Add mutex locking for many-reader/1-writer

khwilliamson opened the pull request:
The mutex macros already in perl are sufficient to allow us to emulate
this type of locking, which may also be available natively, but I don't
think it is worth the effort to use the native calls.

ilmari commented on thread.h:
```diff
@@ -144,6 +144,52 @@
+
+#  define PERL_WRITE_UNLOCK(mutex)  MUTEX_UNLOCK(&mutex)
+
+#define THREAD_CREATE(thr, f)	(thr->self = cthread_fork(f, thr), 0)
+#define THREAD_POST_CREATE(thr)	NOOP
```
These 5005threads remnants were removed in commit 573edd37603ddf5d0137224a22fa2ec4915388fe, please don't re-add them.

khwilliamson replied:
I thought I had gotten rid of those (:

Leont commented:
It's probably helpful to document that this is a Read-preferring implementation.

ilmari commented:
Is there an actual planned use case for this feature? I don't think we should be adding APIs just because they might be useful.

khwilliamson commented:
Of course.

There is an exclusive lock in 5.32 that would be better being a writer-only lock.  ENV_LOCALE_READ_LOCK.  And there are places where a plain ENV or plain LOCALE many-reader lock could be used.  There are a number of libc functions whose behaviour is undefined if one thread is writing while others are reading, but the behaviour is well-defined if multiple readers are doing so at the same time.  Perhaps the name of these locks should be changed to, say WRITE or MANY_READER, instead of READ.  Suggestions welcome

tonycoz commented:
If the mutex, counter and condition object are meant to be used together they should be in a struct, preferably with a typedef.

This would also simplify using a native implementation if available and faster.

tonycoz reviewed (changes requested):
per my comment in the conversation tab

khwilliamson committed:
Add mutex locking for many-reader/1-writer

The mutex macros already in perl are sufficient to allow us to emulate
this type of locking, which may also be available natively, but I don't
think it is worth the effort to use the native calls.

tonycoz commented:
This looks fine, hopefully there will be an API to initialize the composite by the time it's used.
